### DISCUSSION TITLE
Bulk rule creation? (Want to efficiently categorize all existing transactions, but also future ones)

### DISCUSSION THREAD

**Discussion by binarybumpkin** [firefly-iii/firefly-iii#6236] [Firefly III]
Firstly, James (@JC5):

Thanks for such a great product, for all the excellent documentation, and your engagement in the GitHub issues, forums, discussions, etc. for such a contribution to the community. 

**Background**
I've only just discovered FireFly 3/III and the Data Importer, which I've gotten running with docker-compose, behind swag as an nginx reverse proxy with Let'sEncrypt SSL certificates. I've managed to get the Data Importer running and importing from CSV, and had a good read through the docs, on understanding the account flow, and started bringing in transactions then setting account balances for Asset accounts (including a credit card). Main goal is first to see if I can download CSVs from all my banks periodically to get "up-to-date", then set some budgetary targets. Next goal is to extend and automate the generation of the CSVs using python/selenium or with `boobank` that I discovered in another thread, so I can monitor the goals more easily, instead of realising later that I've blown my targets after downloading/importing CSVs. For anyone reading this and wanting to do the same, GitHub Issue #557 (https://github.com/firefly-iii/firefly-iii/issues/557) was such a delight to read, and I recommend it to many.

**Current Task**
Like #4328 (https://github.com/firefly-iii/firefly-iii/issues/4238) for bulk editing transactions, I'm trying to categorize everything - but I don't just want to do it for 000's of transactions once, then have to keep doing it. My goal is to create rules as I go for transactions where I can clearly see unique keywords or phrases in the description for transactions I'm likely to make again (e.g. fast food purchases, etc.) where the Category can be very easily mapped.

I've looked at the ability to create rules, but like #4328, it's cumbersome to do in bulk - even creating rules from transactions, it's necessary to remove all the additional filter triggers that aren't necessary or would make it less generic, etc. and switching back and forth between screens takes time. I've been managing these sorts of things in spreadsheets previously, and building up known vendors/category pairs, etc. to map common purchases to a budget category like this, but I want to move away from spreadsheets to FireFly completely, if I can. 

Which brings me to my question that I haven't been able to find in the docs, forums, discussions, or issues.

**Question**
 Is there any way to bulk create rules for categorizing transactions? I'd like to rapidly categorize transactions and monitor my budget goals during the month by bulk creating rules to match transaction descriptions and assign a category to them.

If this isn't a current feature or easy to do and is more of a feature request, are there technical ways it could be done (e.g. via SQL inserts) or a script of some sort? 

Open to suggestions on how to accomplish that, and happy to contribute back any discoveries/solutions I come up with.

**TLDR**
FireFly 3 is awesome, Data Importer via CSV is excellent, I can import CSVs in bulk and will eventually write scripts for auto-importing transactions from CSV via the data importer, but I want to bulk create rules which match transaction descriptions so I can more efficiently and rapidly categorize transactions and monitor my budget goals during the month - happy for any technical solutions, e.g. SQL inserts, or ways to template/script things in the meantime if it's a feature request, just not sure if it's possible already.

Thanks!

**Comment by JC5** [comment]
Thanks for the compliments!

The new release (still a few months away though) will have pre-storage webhooks. Until then, you can either use normal webhooks or use the API to update your transactions (in combination with webhooks).

Webhooks are currently only available over the API.

https://api-docs.firefly-iii.org/#/webhooks/storeWebhook

Using webhooks, you could:

- Catch the submission of any transaction.
- Extract the group ID and the necessary data.
- Submit an update routine to the API.

You could also do this using rules (they also have API endpoints) but webhooks may give you the flexibility to program something locally.

**Reply by binarybumpkin** [reply]
Thanks, @JC5 - I hadn't seen the API docs yet, so that's a nice discovery.

I'm not sure how the webhooks are relevant, but I did see the API to create and store a new rule:
https://api-docs.firefly-iii.org/#/rules/storeRule

As I've probably got most of what I need in spreadsheets already (e.g. JOES PIZZERIA as a unique phrase that I know could be used as a rule to match transactions which contain that in the description, then match it to the Budget Category for Dining out) I could probably write a script to take two arguments, "Description contains phrase ABC" and "Assign to Category XYZ", then run a for loop on the pairs I've got in a CSV, making calls to the FireFly storeRule API to create the rules in bulk.

The rules still need to be triggered, so I could re-import the transactions, or add a call in to fire each rule after creation in the same for loop:
https://api-docs.firefly-iii.org/#/rules/fireRule 

Once the rules are created and applied, then I'll categorize the remaining transactions manually, and it's "up-to-date". Only need to create more rules if new transactions come in that are likely to come in again, or categorize transactions manually, but there'll be fewer to deal with once the backlog is imported and processed.

Just need to work out how to authorize using the API/token/etc. and should have something workable, I think!

**Reply by JC5** [reply]
Alright, let me know if you need help!

**Reply by binarybumpkin** [reply]
Thanks - this looks like a pretty good place to start!
https://github.com/ms32035/firefly-iii-client

**Reply by binarybumpkin** [reply]
Python script here is basically written for it:
https://github.com/ms32035/firefly-iii-client/blob/master/docs/RulesApi.md#store_rule
